[Chloe Lim Xinying] iP

chloeelim opened the pull request:
### Duke

> "Hello from Duke" - Duke [(Don't click me)](https://www.youtube.com/watch?v=dQw4w9WgXcQ)

Duke can do ~~many~~ things! Like:

- say **hello** to you 👋 
- say _goodbye_ to you (just tell duke `bye`) 👋 👋 
- keep track of your tasks ✅  (just tell duke:
1. `todo <this is my to-do task description>`
2. `deadline <this is my deadline task description> /by DD/MM/YYYY HH:MM`
3. `event <this is my event task description> /at DD/MM/YYYY HH:MM`
- save your tasks locally 💾 

You can even mark your tasks done with the command `mark <task number>`, just like any regular checklist:

- [ ] this is my to-do task description
- [ ] this is my deadline task description (by: Tuesday, 30 Aug 2022 11:18)
- [ ] this is my event task description (at: Tuesday, 30 Aug 2022 11:19)

---
If you Java programmer, you can use it to practice Java too. Here's the `main` method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

chloeelim committed:
Level 1. Greet, Echo, Exit

Skeletal implementation of Duke that:
* greets the user,
* echoes commands entered by the user,
* and exits when the user types bye

chloeelim committed:
Level 2. Add, List

Adds ability to store text entered by the user, and return them in a numbered
list when requested.

chloeelim committed:
Level 3. Mark as Done

Adds the ability to mark tasks as done/ undone. Also refactors out
tasks into its own class.

chloeelim committed:
Level 4. ToDos, Events, Deadlines

Add support for tracking three types of tasks: To Dos, Events and Deadlines.

chloeelim committed:
Automated Text UI Testing

Use the input/output redirection technique to semi-automate the testing of Duke.

chloeelim committed:
Level 5. Handle Errors

Added custom exception to handle Duke exceptions.

chloeelim committed:
Level 6. Delete

Adds functionality for task deletion from the task list.

chloeelim committed:
A-Enums: Use Enums

Refactor code to use enums, as well as refactor out task list into its own class.

chloeelim committed:
Add JavaDocs

Adds JavaDocs to TaskType and TaskList

chloeelim committed:
Level 7. Save

Adds functionality to save tasks to harddrive and load saved tasks at the start
of the program, if there were any previously saved tasks.

chloeelim committed:
Level 8. Dates and Times

Adds datetime recognition for deadline and event tasks in dd/MM/yyyy or dd/MM/yyyy HH:mm format.

chloeelim committed:
Merge branch 'branch-Level-7'

chloeelim committed:
Merge branch 'branch-Level-8'

chloeelim committed:
fix: fix errors introduced from merging the 2 branches

chloeelim committed:
Use More OOP

Refactor out user interactions, parsing, storage, and commands.

chloeelim committed:
Organize into Packages

Divides classes into respective packages (duke, duke.task, duke.command, duke.exception).

chloeelim committed:
Merge branch 'add-gradle-support' of https://github.com/nus-cs2103-AY2223S1/ip

chloeelim committed:
fix and set up checkstyle

chloeelim committed:
Add JUnit Tests

Adds JUnit Tests to test ToDo and TaskList class.

chloeelim committed:
A-JavaDoc

Add JavaDoc comments to the code.

chloeelim committed:
A-CodingStandard

Tweaks the code to comply with a given coding standard.

chloeelim committed:
Level 9. Find

Give users a way to find a task by searching for a keyword.

chloeelim committed:
Merge branch 'branch-A-JavaDoc'

chloeelim committed:
Merge branch 'branch-A-CodingStandard'

chloeelim committed:
Merge branch 'branch-Level-9'

chloeelim committed:
fix: infinite loop

chloeelim committed:
refactor: fix checkstyle errors

chloeelim committed:
Level 10. GUI
Add a GUI to Duke. Use the JavaFX technology to implement the GUI.

chloeelim committed:
Merge branch 'branch-Level-10'

chloeelim committed:
feat: add support for event start-end with datetime

adds functionality to support start and end datetime for events, also
adds error handling for unparsable saved task formats (ignores these tasks).

chloeelim committed:
feat: add event start-end datetime validation

adds datetime validation for events to ensure start is not after end datetime.

Puakii commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,58 @@
+    /**
+     * Constructs a storage to be managed at the specified file location.
+     *
+     * @param directoryPath The path of the directory containing the file
+     * @param filename The filename of the file to save the data to
```
Perhaps you should add a full stop after every parameter description (see Java Coding Standard).

Puakii commented on src/main/java/duke/task/Event.java:
```diff
@@ -0,0 +1,64 @@
+    }
+
+    /**
+     * Parses the Event into a savable string format, ready to be written to the hard disk.
+     * @return Savable string representation of the Deadline Task
```
I think you forgot to leave a line here.

Puakii reviewed (commented):
Really impressed with your code quality! Name of variables and methods are clearly defined and coding standard are almost free from mistakes. Could do a tiny touch-up and it would be perfect.

Puakii commented on src/test/java/TaskListTest.java:
```diff
@@ -0,0 +1,39 @@
+
+public class TaskListTest {
+    @Test
+    public void accessTask_accessExistentTask_success() throws DukeException {
```
Great method naming!

rgonslayer commented on src/main/java/duke/command/AddTaskCommand.java:
```diff
@@ -0,0 +1,47 @@
+package duke.command;
```
Good job organising your commands into classes and placing it into a package

rgonslayer reviewed (commented):
Generally well done! No comments from myself as the other reviewer has already pointed out most of the things to approve!

Yingming23 commented on src/main/java/duke/gui/DialogBox.java:
```diff
@@ -0,0 +1,63 @@
+package duke.gui;
```
I like the use of packages for all the GUI classes.

Yingming23 commented on src/main/java/duke/task/Event.java:
```diff
@@ -0,0 +1,64 @@
+/**
+ * An event is a task that starts at a specific time and ends at a specific time.
+ */
+public class Event extends Task {
```
I like the extra work done to put an end date for Event.

Yingming23 reviewed (commented):
Overall, your code looks really clean and the OOP is done really well. I like all the extra effort put into the extra functions.

chloeelim committed:
feat: improve gui (list)

improves gui by adding topbar, changing icons, and creating a list view.

chloeelim committed:
feat: add more personality and fix list view styling

adds more personality to duke (renamed to bobo), and fixes some issues
with view (specifically for tasks with long descriptions). also adds list
view to other commands (addtask, deletetask, findtask, mark/ unmark task).
also fixes alignment of datesection (offset to the tasknumber).

chloeelim committed:
fix: fix minor style issues and stringified message for number of tasks

chloeelim committed:
Style: fix style errors and improve code quality

Code has a few style violations, such as missing punctuation at
the end of the params in the JavaDocs.

The code quality could also be improve, by enforcing the SLAP principle,
using guard clauses and improving readability.

Let's fix all the styling errors and try to ensure that code quality adheres
to the coding standard.

chloeelim committed:
A-Assertions. Use Assertions

The code does not use assertions.

Assertions define assumptions about the program state so that the runtime can verify them.
Adding assertions can help us catch bugs by verifying our assumptions of the state
of the program.

Let's add assertions at some critical junctures of the code and verify that the runtime state
of our program coheres with our assumptions of the state of the program.

This will allow us to verify the correctness of the program, and can help us catch bugs
in the program.

chloeelim committed:
A-Streams. Use Streams

chloeelim committed:
Merge pull request #1 from chloeelim/branch-A-CodeQuality

Style: fix style errors and improve code quality

chloeelim committed:
Merge branch 'master' into branch-A-Assertions

chloeelim committed:
Merge branch 'master' into branch-A-Streams

chloeelim committed:
Merge pull request #2 from chloeelim/branch-A-Assertions

A-Assertions. Use Assertions

chloeelim committed:
Merge branch 'master' into branch-A-Streams

chloeelim committed:
Merge pull request #3 from chloeelim/branch-A-Streams

A-Streams. Use Streams

chloeelim committed:
A-CI. Set up CI

chloeelim committed:
C-NaturalDates. adds more flexibility to datetime parsing

bobo only accepts a very limited and rigid set of datetime formats (dd/MM/YYYY or
dd/MM/YYYY HH:mm).

The reference of dates in bobo is thus, very inflexible and not very user friendly.

Let's add a new class, DateTimeParse to handle datetime parsing. We can use
a linked hashmap to store a large variety of date or datetime formats. This hashmap
stores the regex expression corresponding to the date or datetime pattern so that
we can look through this hashmap for a matching regex expression to the datetime string,
and map it to its corresponding datetime formatter. Finally, we can parse the datetime
string into a datetime object with the formatter.

Abstracting out datetime parsing into its separate class helps enforce OOP principles
and readability of the code. Especially since it gets rather lengthy and verbose.
Abstracting it out will also make datetime parsing also more extensible so that
it is easier to add new datetime formats to be supported in the future.

However, the code could be further refactored so it is cleaner- such as using enums
and abstracting out some of the string processing into its own separate methods.

chloeelim committed:
Merge pull request #4 from chloeelim/branch-C-NaturalDates

C-NaturalDates. adds more flexibility to datetime parsing

bobo only accepts a very limited and rigid set of datetime formats (dd/MM/YYYY or dd/MM/YYYY HH:mm).

The reference of dates in bobo is thus, very inflexible and not very user friendly.

Let's add a new class, DateTimeParse to handle datetime parsing. We can use a linked hashmap to store a large variety of date or datetime formats. This hashmap stores the regex expression corresponding to the date or datetime pattern so that we can look through this hashmap for a matching regex expression to the datetime string, and map it to its corresponding datetime formatter. Finally, we can parse the datetime string into a datetime object with the formatter.

Abstracting out datetime parsing into its separate class helps enforce OOP principles and readability of the code. Especially since it gets rather lengthy and verbose. Abstracting it out will also make datetime parsing also more extensible so that it is easier to add new datetime formats to be supported in the future.

However, the code could be further refactored so it is cleaner- such as using enums and abstracting out some of the string processing into its own separate methods.

chloeelim committed:
refactor: numerical date format

refactors numerical date format into its own class for extensibility.

chloeelim committed:
feat: add more datetime supported formats

adds more datetime supported formats:
  * support 2, 3 and 4 digit numerical date formats that may
    be missing at least one of the date (day/ month/ year) fields
  * support text-dates (month field) like 11 Oct 2022, August 2020,
    22 March
  * support 12-hour time formats (11:23 PM, 2:00am...)

also adds a parent class for dateformats to abstract out common methods.
to look into also abstracting out time into its own separate class for
extensibility and to adhere to oop-principles as it is getting a little
unwieldy.

chloeelim committed:
style: fix checkstyle errors

rectifies checkstyle errors found in the code.

chloeelim committed:
feat: support natural dates and more time formats

Adds support for natural dates (like today, tomorrow...) and also
extends support for time formats:
  * Now, dates with only time are also possible (date will be
    automatically set to today).
  * In addition, 12-hour time format is now also supported (with AM/PM)

To do this, abstract natural dates and time into their own classes
to keep in line with OOP principles.

chloeelim committed:
feat: add support for more natural dates

Adds support for more natural dates (days of the week- such that
{day of the week} will be interpreted to next {day of the week}).

chloeelim committed:
C-Update. adds update functionality

After creating tasks, you cannot update (ie: edit) them directly.

To edit a task, a user would have to first delete the task, then
create a new task with the updated details.

Let's simplify this process by allowing the user to directly update
an existing task. The update function will edit the task details
according to user's inputted details. It will not delete and create
a new task so that the order of the task in the task list remains.

However, the command to update a task might be a little troublesome
since the user would have to edit all the given fields (eg: even
if the user only wants to edit the end date of an event, the user
would have to re-enter the event details and start time as well).
Future commits may want to look into improving this aspect.

chloeelim committed:
style: fix checkstyle errors

fixes checkstyle errors in the code by removing unnecessary imports.

chloeelim committed:
Merge pull request #6 from chloeelim/branch-C-Update

C-Update. Add Update (Edit) Functionality for Tasks

chloeelim committed:
B-ViewSchedules: add feature to view schedules

Adds feature to view schedule, given a specified date or datetime.
Also fixes some minor bugs with datetime parsing (4 digit numeric
should try to parse as a date before time), as well as some missing
JavaDocs in NaturalDate.

chloeelim committed:
Merge branch 'branch-B-ViewSchedules'

Merges in the feature to view schedules on a specified date.

chloeelim committed:
feat: add completion dates

adds feature to track the completion datetime of each task (when it
was marked completed). this info is displayed with a tooltip
when hovering over a checked box.

chloeelim committed:
A-Personality. update bobo icons

use better (higher resolution) and more expressive icons.

chloeelim committed:
C-Statistics: visualise statistics

adds 'stats' command to visualise breakdown of all completed tasks
by tasktype with a pie chart.

chloeelim committed:
feat. update gui

updates pie chart gui to include breakdown numbers of each task,
and updates bobo icon.

chloeelim committed:
Style: fix checkstyle errors

Fixes checkstyle errors.

chloeelim committed:
Merge pull request #7 from chloeelim/branch-C-Statistics

Branch c statistics

chloeelim committed:
A-UserGuide. add user guide

Adds user guide to the app.

chloeelim committed:
Fix README

Fixes broken image in README file.

chloeelim committed:
Style: minor edits to user guide

Add details about schedule feature, and make some minor edits to the formatting.

chloeelim committed:
Feat: add new statistics (bar chart)

Adds ability to generate bar chart statistics for day/ week.

chloeelim committed:
Doc: Update README with new commands

Updates README with the bar chart commands.

chloeelim committed:
Fix: bar chart type statistic command more flexibility

Allows users to enter bar chart type statistic commands without declaring
an explicit type flag (/day or /week) while specifying a date. In other words,
a command like `stats -b yesterday` will now be interpreted as generating
a bar chart daily statistic for yesterday.

chloeelim committed:
Fix: user guide broken images

fixes missing images in user guide.

chloeelim committed:
Merge branch 'branch-C-Statistics'

chloeelim committed:
Test: Add more unit tests

Adds unit tests for most of the classes.

chloeelim committed:
Fix: change xmlns JavaFX version

chloeelim committed:
Docs: Update README

Update README feature with stats (Bar chart).
